### PR TITLE
Make HealthCheckedEndpointGroup gracefully update new endpoints

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

We have added a lot of workarounds and fixed bugs for
the race condition and timing issue for `HealthCheckedEndpiontGroup`.
https://github.com/line/armeria/blob/43bd08248ba86ab5942f781da33b11a42fe3b1a5/core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java#L145-L161
However, we still got a bug report from `HealthCheckedEndpointGroup`.
https://github.com/line/armeria/issues/3173#issuecomment-953826146

I think another workaround could not prevent new bugs from edge cages.
We should reconsider the current design of `HealthCheckedEndpointGroup`
implmentation.

Design note:

If an endpoint group is added, we have to destroy the old endpoint group
after the first health check for the new endpoint group is completed.
Explicit context groups to distinguish the new endpoints with
the old endpoints should be useful to manage the lifecycle of endpoints 
and health checker contexts.

For example, if endpoints `[1, 2, 3]` are added, we can allocate them to `context1`.
If new endpoints `[3, 4, 5]` are added, they will be allocated to `context2`.
```
context1[1(checked), 2(checked), 3(checked)] -> context2[3(checked), 4(unchecked), 5(unchecked)]
```
Once all endpoints for context2 is checked, we can safely destroy
context1.
```
context1[deleted] -> context2[3(checked), 4(checked), 5(checked)]
```

Modifications:

- Add `HealthCheckContextGroup` to cluster a set of endpoints set together.
- Remove various workarounds in the constructor of
  `HealthCheckedEndpointGroup` is removed.
  - The root causes of the workarounds are fixed by the new design.
- Manage the lifecycle of an endpoint group using `contextGroupChain` in
  `HealthCheckedEndpointGroup`.
  - A new contexts group is fully initialized, old contexts will be destroyed.
- Move `DefaultHealthCheckerContext` to the top level.
- Revamp `HealthCheckStrategy` .
  - Remove internal states from `PartialHealthCheckStrategy`.
    - `PartialHealthCheckStrategy` maintains its own healthy and unhealthy
      endpoints that can delegate to `HealthCheckedEndpointGroup`.
  - Simplify the API and implementation.
- Remove blocking operations from the constructor of `HealthCheckedEndpointGroup`.
  - Note that you should use `HealthCheckedEndpointGroup.whenReady()` to
     wait for or the initial endpoint groups.

Result:

- Fix https://github.com/line/armeria/issues/3173#issuecomment-953826146 reported by @imasahiro
- Clean up `HealthCheckedEndpointGroup` implementation
- You no longer see an `EmptyEndpointGroupException` when the entire
  endpoints are changed in the `HealthCheckedEndpointGroup`

### CODE REVIEW

**Review by minwoox — COMMENTED** [review]
Thanks! This looks great! 😄
Left some minor comments. 😉

File: core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java [hunk 1]
```diff
@@ -140,113 +133,124 @@ public static HealthCheckedEndpointGroupBuilder builder(EndpointGroup delegate,
...
-        healthCheckStrategy.updateCandidates(candidates);
-        refreshContexts();
-    }
+    private void setCandidates(List<Endpoint> candidates) {
+        final List<Endpoint> endpoints = healthCheckStrategy.select(candidates);
+        final HashMap<Endpoint, DefaultHealthCheckerContext> contexts = new HashMap<>(endpoints.size());
+        for (Endpoint endpoint : endpoints) {
+            final DefaultHealthCheckerContext context = findContext(endpoint);
+            if (context != null) {
+                contexts.put(endpoint, context.retain());
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java:152]: Oops... Indeed. 😱

File: core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java [hunk 2]
```diff
@@ -140,113 +133,124 @@ public static HealthCheckedEndpointGroupBuilder builder(EndpointGroup delegate,
...
-                                            ctx.endpoint(), cause);
-                                return null;
-                            }))
-                            .toArray(CompletableFuture[]::new));
-            contexts.clear();
+        synchronized (contextGroupChain) {
+            final ImmutableList.Builder<CompletableFuture<?>> completionFutures = ImmutableList.builder();
+            for (HealthCheckContextGroup group : contextGroupChain) {
+                for (DefaultHealthCheckerContext context : group.contexts().values()) {
+                    final CompletableFuture<?> closeFuture = context.release();
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HealthCheckedEndpointGroup.java]: It should not throw any exception if we correctly implement the logic. Let me use `assert` for the verification.

**Review by minwoox — APPROVED** [review]
Thanks!
Hope not to see the `EmptyEndpointException` anymore. 🤣

**Comment by minwoox** [comment]
Could you check the CI?

**Comment by ikhoon** [comment]
> Could you check the CI?

Oops...

**Comment by ikhoon** [comment]
Fixed by [`0e956c7` (#3927)](https://github.com/line/armeria/pull/3927/commits/0e956c75b41d561006ce93e9e6b703f44540bd1a)

**Pull request merged**